helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/FaultZoneAwareConstraint.java

```diff
@@ -26,27 +26,27 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import static org.apache.helix.controller.rebalancer.waged.constraints.HardConstraint.ValidationResult.fail;
 
 class FaultZoneAwareConstraint extends HardConstraint {
 
   private static final Logger LOG = LoggerFactory.getLogger(FaultZoneAwareConstraint.class);
 
   @Override
-  boolean isAssignmentValid(AssignableNode node, AssignableReplica replica,
+  ValidationResult isAssignmentValid(AssignableNode node, AssignableReplica replica,
```

Comment:
We do not allowed CHANGE api.

Reply:
Yes, this is a very valid point. 

Another approach is that I can expose another function that fetches the error detail if there is any validation error (keeping current behavior as it is). But in order to do that, we might have to duplicate the logic of `isAssignmentValid`, Or, create a common function that spits out error message or boolean, depending upon how we are accessing it.